libsbml 5.20.2 (new formula)

Waschina opened the pull request:
libsbml is a library for handling text files in Systems Biology Markup Language (SBML).

A brief explanation for the additional configuration flags (-DENABLE_FBC=ON -DENABLE_GROUPS=ON):

These enable two optional extensions for the libsbml library, namely "fbc" and "groups". Both are relevant for the research community that uses linear programming to predict metabolic processes in living cells (i.e., flux balance analysis). apt and dnf linux builds also include these two extensions.

- [x] Have you followed the [guidelines for contributing](https://github.com/brewsci/homebrew-bio/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/brewsci/homebrew-bio/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source FORMULA`, where `FORMULA` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict FORMULA` (after doing `brew install FORMULA`)?

-----

This is my first formula. Any feedback is welcome.

Waschina commented:
Thank you for approving the CI workflows.

```
2024-04-25T01:11:02.0384350Z /opt/homebrew/share/cmake/Modules/FindZLIB.cmake -> /opt/homebrew/Cellar/cmake/3.29.2/share/cmake/Modules/FindZLIB.cmake
```

There seems to be a conflict with the cmake module `FindZLIB.` cmake itself provides [a module under this name](https://github.com/Kitware/CMake/blob/master/Modules/FindZLIB.cmake), and libSBML comes with [its own module](https://github.com/sbmlteam/libsbml/blob/development/CMakeModules/FindZLIB.cmake) using the same name.

Does anyone know how to handle such a conflict in a formula?

YoshitakaMo commented:
Thank you for submitting a new formula! 
Please test this:
```ruby
  depends_on "check" => :build
  depends_on "cmake" => :build
  depends_on "pkg-config" => :build
  depends_on "swig" => :build

  uses_from_macos "bzip2"
  uses_from_macos "libxml2"
  uses_from_macos "zlib"

  def install
    # avoid an error "invalid conversion from ‘const xmlError*’"
    ENV.append_to_cflags "-fpermissive" if OS.linux?
    args = %w[
      -DWITH_SWIG=OFF
      -DWITH_ZLIB=OFF
      -DWITH_BZIP2=OFF
      -DENABLE_COMP=ON
      -DENABLE_FBC=ON
      -DENABLE_GROUPS=ON
      -DENABLE_L3V2EXTENDEDMATH=ON
      -DENABLE_LAYOUT=ON
      -DENABLE_MULTI=ON
      -DENABLE_QUAL=ON
      -DENABLE_RENDER=ON
    ]
    args << "-DLIBSBML_DEPENDENCY_DIR=#{HOMEBREW_PREFIX}"
    system "cmake", "-S", ".", "-B", "build", *args, *std_cmake_args
    system "cmake", "--build", "build"
    system "cmake", "--install", "build"
  end
```

- `-DWITH_BZIP2=OFF` and `-DWITH_ZLIB=OFF` are required to solve the conflict.
- `-DLIBSBML_DEPENDENCY_DIR=#{HOMEBREW_PREFIX}` is required to use Homebrew's libxml2 (for Linux).
- Appending CFLAGS="-fpermissive" was required to compile on Linux.
- Other arguments were copied from [Bioconda's recipes](https://github.com/bioconda/bioconda-recipes/blob/master/recipes/libsbml/build.sh)

Waschina commented:
Thank you, @YoshitakaMo, for the suggestion and for explaining the changes.

I made only minor modifications to your suggested formula:
- re-included bzip2 since this dependency seems to be without conflict
- removed `uses_from_macos "zlib"` since zlib is disabled as a dependency of libsbml with the `-DWITH_ZLIB=OFF` flag
- removed swig as build dependency since it is turned off with  `-DWITH_SWIG=OFF`.

YoshitakaMo commented:
I guess `args << "-DLIBSBML_DEPENDENCY_DIR=#{HOMEBREW_PREFIX}"` is required to pass the CI.

Waschina commented:
Sorry, that I missed that line. It's now committed

YoshitakaMo commented:
Merged! Thank you for your PR!

Waschina commented:
Awesome! Thank you for being so helpful @YoshitakaMo
